workflow/controller/operator_data_source_processor.go

```diff
@@ -0,0 +1,17 @@
+package controller
+
+import (
+	"fmt"
+
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+)
+
+type operatorDataSourceProcessor struct{}
```

Comment:
what is this for?

Reply:
The executor and the operator use the same code to apply a transformation (defined in `workflow/data/data.go`). When the executor uses it, it needs to pass in some functions that only it possesses (namely `ProcessArtifactPaths`), so it passes it by implementing the `DataSourceProcessor` interface (defined in `data_types.go`). Since the executor cannot source artifact paths, it passes its own null implementation of the interface. The error should never get triggered, because if we're processing the data template in the controller, we know it won't source artifact paths.